src/pycrunch/elements.py

```diff
@@ -277,12 +277,24 @@ class ElementResponseHandler(lemonpy.ResponseHandler):
 
     parsers = {"application/json": parse_json_element_from_response}
 
+    def auth_credentials(self):
+        """
+        Returns the credentials used to auth this session.
+        """
+        if self.session.token:
+            creds = {"token": self.session.token}
+        else:
+            creds = {"email": self.session.email, "password": self.session.password}
+        return creds
+
     def status_401(self, r):
         login_url = r.json()["urls"]["login_url"]
         if r.request.url == login_url:
+            # This means that the requests was made on the login_url, so it was
+            # an authentication attempt. It means the login failed.
             raise ValueError("Log in was not successful.")
 
-        creds = {"email": self.session.email, "password": self.session.password}
+        creds = self.auth_credentials()
```

Comment:
I think the problem here is that attempting to login and store the cookie is only applicable with username/password authn. If the user gets a 401 with an apikey then the key they have stored doesn't work anymore so rather than sending the apikey to the login endpoint which isn't supported we probably want to raise an exception that the token they are attempt to use isn't valid.

Reply:
Thanks for this clarification. This code is performing a 2nd attempt to log in if the firs request failed. I am 100% to not do this for api keys. And leave the older behavior only for email/password.